src/celeritas/optical/interactor/WavelengthShiftInteractor.hh

```diff
@@ -0,0 +1,174 @@
+ * primary optical photon.
+ *
+ * \note This performs the same sampling routine as in the G4OpWLS class of
+ * the Geant4 release 11.2.
+ */
+class WavelengthShiftInteractor
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using Energy = units::MevEnergy;
+    using ParamsRef = NativeCRef<WavelengthShiftData>;
+    using SecondaryAllocator = StackAllocator<TrackInitializer>;
+    //!@}
+
+  public:
+    // Construct with shared and state data
+    inline CELER_FUNCTION
+    WavelengthShiftInteractor(ParamsRef const& shared,
+                              ParticleTrackView const& particle,
+                              OpticalMaterialId const& mat_id,
+                              SecondaryAllocator& allocate);
+
+    // Sample an interaction with the given RNG
+    template<class Engine>
+    inline CELER_FUNCTION Interaction operator()(Engine& rng);
+
+  private:
+    //// DATA ////
+
+    // Incident photon energy
+    Energy const inc_energy_;
+    // Sample distributions
+    PoissonDistribution<real_type> sample_num_photons_;
+    ExponentialDistribution<real_type> sample_time_;
+    // Grid calculators
+    GenericCalculator calc_energy_;
+    GenericCalculator calc_cdf_;
+    // Allocate space for secondary particles
+    SecondaryAllocator& allocate_;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with shared and state data.
+ */
+CELER_FUNCTION
+WavelengthShiftInteractor::WavelengthShiftInteractor(
+    ParamsRef const& shared,
+    ParticleTrackView const& particle,
+    OpticalMaterialId const& mat_id,
+    SecondaryAllocator& allocate)
+    : inc_energy_(particle.energy())
+    , sample_num_photons_(shared.wls_record[mat_id].mean_num_photons)
+    , sample_time_(real_type{1} / shared.wls_record[mat_id].time_constant)
+    , calc_energy_(shared.energy_cdf[mat_id], shared.reals)
+    , calc_cdf_(GenericCalculator::from_inverse(shared.energy_cdf[mat_id],
+                                                shared.reals))
+    , allocate_(allocate)
+{
+    CELER_EXPECT(inc_energy_.value() > calc_energy_(0));
+}
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Sampling the wavelength shift (WLS) photons as in G4OpWLS of the Geant4
+ * release 11.2 with a modification in sampling the reemission spectrum.
+ */
+template<class Engine>
+CELER_FUNCTION Interaction WavelengthShiftInteractor::operator()(Engine& rng)
+{
+    // Sample the number of photons generated from WLS
+    unsigned int num_photons = sample_num_photons_(rng);
+
+    if (num_photons == 0)
+    {
+        // Return unchanged photon without reemitted optical photons
+        return Interaction::from_unchanged();
```

Comment:
Yes, it won't be actually absorbed (i.e., no wavelength shifting is applied) even though the optical photon should be moved for the proposed length - in principle, the zero optical photon may be emitted from the excited WLS material (something like the internal conversion which may not be allowed in the selection rule, but can happen quantum mechanically).

Reply:
This still sounds wrong: even though we've sampled a photon interaction at this point, no interaction takes place? The original photon continues with its energy, direction, polarization?

Reply:
Okay, I simply changed the action from `from_unchanged` to `from_absorption` (i.e, kill the primary optical photon) so that the outer loop treats WLS as same as the absorption process - this is equivalent to what Geant4 does.